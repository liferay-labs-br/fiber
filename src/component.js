import createComponent from './component-lifecycle';

class Component {
	
  /**
   * Constructor
   * isMounted prevents the component call the componentWillMount() function
   * when the user use the constructor in his Fiber component
   */
  constructor(props, context, isMounted) {
    this.props = props;
    this.context = context;

    if (isMounted) {
      this.componentWillMount();
    }
  }

  /**
   * Lifecycle
   * Corresponds created()
   */
  componentWillMount() {}

  /**
   * Lifecycle
   * Corresponds disposed()
   */
  componentWillUnmount() {}

  /**
   * Lifecycle
   * Corresponds shouldUpdate()
   */
  shouldComponentUpdate() {}

  /**
   * Lifecycle
   * Corresponds rendered()
   */
  componentDidUpdate() {}

  setState(state, callback) {}

  render() {}
}

const buildComponentFromVNode = (vnode, context) => {
  let props = Object.assign({}, vnode.attributes);
  return createComponent(vnode.nodeName, props, context);
};

export {
  Component,
  buildComponentFromVNode,
};
