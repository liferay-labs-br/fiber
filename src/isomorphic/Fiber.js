/**
 * Copyright (c) 2017-present, Fiber.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import Component from './FiberComponent';
import createElement from './FiberElement';
import renderFactory from './FiberRender';

export default {
  Component,
  createElement,
  renderFactory
};

export {
  Component,
  createElement,
  renderFactory
};
