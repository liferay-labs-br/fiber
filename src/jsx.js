export function jsx(nodeName, attributes, ...args) {
  let children = args.length ? [].concat(...args) : null;

  return { nodeName, attributes, children };
}
