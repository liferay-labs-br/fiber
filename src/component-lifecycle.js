import { Component } from './component';

export default function createComponent(Constructor, props, context) {
  let instance;

  if (Constructor.prototype && Constructor.prototype.render) {
    instance = new Constructor(props, context);
    Component.call(instance, props, context, true);
    instance = instance.render();
  } else {
    instance = Constructor(props);
  }

  return instance;
}
