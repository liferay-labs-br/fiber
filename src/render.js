import { buildComponentFromVNode } from './component.js';
import renderUtils from './utils/render-utils';

export default function render(vnode, parent) {
	if (vnode == null || typeof vnode === 'boolean') vnode = '';

	if (typeof vnode==='string') return document.createTextNode(vnode);
	
	if (typeof vnode === 'function') vnode = vnode();

	if (typeof vnode.nodeName === 'function') vnode = buildComponentFromVNode(vnode, {});
	
	let node = document.createElement(vnode.nodeName);
	
	let attributes = vnode.attributes || {};
	renderUtils._formatAttributes(node, attributes);
	
	(vnode.children || []).forEach( child => node.appendChild(render(child)) );
	
	if (parent) parent.appendChild(node);
	
	return node;
}
