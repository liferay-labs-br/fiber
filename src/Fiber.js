/**
 * Copyright (c) 2017-present, Fiber.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { Component, createElement } from './isomorphic/Fiber';
import { render } from './renderers/dom/FiberDOM';

export default {
  createElement,
  Component,
  render
}

export {
  createElement,
  Component,
  render
}
