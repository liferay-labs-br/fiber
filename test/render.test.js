import { render, createElement } from '../src/Fiber';

let Foo;

describe('render', () => {
  let artboard;

  beforeAll(() => {
    artboard = document.createElement('div');
		(document.body || document.documentElement).appendChild(artboard);
  });

  beforeEach(() => {
    artboard.innerHTML = '';
  });

  afterAll(() => {
    artboard.parentNode.removeChild(artboard);
		artboard = null;

    if (Foo) {
      Foo.dispose();
    }
  });

  it('should create empty node <div />', () => {
    render(<div />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should create empty node <p />', () => {
    render(<p />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should create empty nodes <foo /> and <bar>', () => {
    render(<foo />, artboard);
    render(<bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should nest empty nodes', () => {
    render((
      <div>
        <p />
        <foo />
        <bar />
      </div>
    ), artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render falsey values', () => {
    render((
      <div>
        {null}{undefined}{false}{0}{NaN}
      </div>
    ), artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render null', () => {
    render(null, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render undefined', () => {
    render(undefined, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render boolean true', () => {
    render(true, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render boolean false', () => {
    render(false, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should render NaN as text content', () => {
    render(NaN, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should render numbers (0) as text content', () => {
    render(0, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should render strings as text content', () => {
    render('Hello world', artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when function return empty', () => {
    let Foo = () => {
      return;
    };

    render(Foo, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when function return bool true', () => {
    let Foo = () => {
      return true;
    };

    render(Foo, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when function return bool false', () => {
    let Foo = () => {
      return false;
    };

    render(Foo, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when function return string', () => {
    let Foo = () => {
      return 'Foo';
    };

    render(Foo, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when function return undefined', () => {
    let Foo = () => {
      return undefined;
    };

    render(Foo, artboard);

    expect(artboard).toMatchSnapshot();
  });
});
