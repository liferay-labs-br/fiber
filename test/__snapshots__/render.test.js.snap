// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should create empty node <div /> 1`] = `
<div>
  <div />
</div>
`;

exports[`render should create empty node <p /> 1`] = `
<div>
  <p />
</div>
`;

exports[`render should create empty nodes <foo /> and <bar> 1`] = `
<div>
  <foo />
  <bar />
</div>
`;

exports[`render should nest empty nodes 1`] = `
<div>
  <div>
    <p />
    <foo />
    <bar />
  </div>
</div>
`;

exports[`render should not render boolean false 1`] = `
<div>
  
</div>
`;

exports[`render should not render boolean true 1`] = `
<div>
  
</div>
`;

exports[`render should not render falsey values 1`] = `
<div>
  <div>
    
    
    
    0
    NaN
  </div>
</div>
`;

exports[`render should not render null 1`] = `
<div>
  
</div>
`;

exports[`render should not render undefined 1`] = `
<div>
  
</div>
`;

exports[`render should not render when function return bool false 1`] = `<div />`;

exports[`render should not render when function return bool true 1`] = `<div />`;

exports[`render should not render when function return empty 1`] = `<div />`;

exports[`render should not render when function return string 1`] = `<div />`;

exports[`render should not render when function return undefined 1`] = `<div />`;

exports[`render should render NaN as text content 1`] = `
<div>
  NaN
</div>
`;

exports[`render should render numbers (0) as text content 1`] = `
<div>
  0
</div>
`;

exports[`render should render strings as text content 1`] = `
<div>
  Hello world
</div>
`;
